### DISCUSSION TITLE
Feature Request: Access Lower-Level .query API for Running Multiple Statements in a Single Query

### DISCUSSION THREAD

**Discussion by renatocron** [prisma/prisma#25349] [Q&A]
### Question

Hello Prisma team,

I have a use case where I need to run multiple SQL statements in a single query, and currently, I am using raw SQL with pg for this. Here is an example of what I am trying to achieve using the pg library:

```typescript
async function updateFunctions(filesToUpdate: { fileName: string; hash: string; content: string }[]) {
    const pool = new Pool({ connectionString: process.env.DATABASE_URL });
    const connection = await pool.connect();

    for (const file of filesToUpdate) {
        try {
            await connection.query('BEGIN');
            await connection.query(file.content);  // Running raw SQL
            await connection.query(
                'INSERT INTO _migrate_pgsql (file_name, hash, last_sql) VALUES ($1, $2, $3) ' +
                'ON CONFLICT (file_name) DO UPDATE SET hash = EXCLUDED.hash, last_sql = EXCLUDED.last_sql, updated_at = CURRENT_TIMESTAMP',
                [file.fileName, file.hash, file.content]
            );
            await connection.query('COMMIT');
        } catch (error) {
            await connection.query('ROLLBACK');
            throw error;
        }
    }
    connection.release();
    await pool.end();
}
```

I would like to achieve the same functionality only Prisma. Specifically, I would like to access a lower-level API where I can:

- Execute raw SQL content from files that can contain multiple SQL statements.
- Use transactions (BEGIN, COMMIT, ROLLBACK) around these operations.
- Run multiple statements in a single query execution.
- 
Currently, Prisma's prisma.$executeRawUnsafe only allows executing one statement per query, which limits my ability to process multiple SQL statements together. I looked at `prisma migrate` but it called the rust API via RPC, so I cannot do the same here.

Is there any way to allow access to a lower-level .query API, similar to pg, even if I need to keep updating the undocumented path to it between version upgrades?

Thank you for your help!

### How to reproduce (optional)

Setup like: 

```ts
async function updateFunctions(filesToUpdate: { fileName: string; hash: string; content: string }[]) {
    await prisma.$transaction(async (tx) => {
        for (const file of filesToUpdate) {
            logger.log(`Updating ${file.fileName}...`);
            try {
                await tx.$executeRawUnsafe(file.content);

                await tx.$executeRaw`
                    INSERT INTO _migrate_pgsql (file_name, hash, last_sql)
                    VALUES (${file.fileName}, ${file.hash}, ${file.content})
                    ON CONFLICT (file_name)
                    DO UPDATE SET hash = EXCLUDED.hash, last_sql = EXCLUDED.last_sql, updated_at = CURRENT_TIMESTAMP;
                `;
            } catch (error) {
                logger.error(`Error updating ${file.fileName}:`, error.message);
                throw error; // Re-throw to trigger transaction rollback
            }
        }
    });
}
```

full script: https://gist.github.com/renatocron/39a1f574286d0472fd9c1a71af875af9 and create a dir `./pgsql` add a file with `select 1; select 2` will thrown 

Raw query failed. Code: `42601`. Message: `ERROR: cannot insert multiple commands into a prepared statement`

**Comment by ludralph** [comment]
HI @renatocron 👋 

Thank you for raising this question.

You can leverage the new [TypedSQL](https://www.prisma.io/docs/orm/prisma-client/using-raw-sql/typedsql) feature in Prisma to solve this issue. While you can't read from SQL files at the moment, you can however create multiple SQL files and generate fully typed functions from them. You can also use transactions within the SQL files for TypedSQL.

Do let us know if this approach would help resolve your issue.

**Reply by renatocron** [reply]
Hi @ludralph 

I recently came across this new feature and haven't tested it yet, but it seems like a good way to handle my views. It would eliminate the need to manually write schemas, except in cases where foreign keys are involved.

Does this mean that TypeSQL can be used for more than just views? For instance, can it handle functions and triggers?

The reason I'm asking is that my .pgsql files don't always consist of simple SELECT statements. I used SELECT as an example, but I also have many complex functions that need to be updated in the database. Managing these functions has been challenging, especially when each change requires a new migration, and ensuring that the most recent migration contains the current version of the procedures is a tedious process.

The script is now fully functional using the pg library, and I've added watch mode as well. If there's a way to integrate this back into Prisma, it would be fantastic, as I plan to use this in multiple projects.

here are the .pgsql files https://github.com/AppCivico/smae/tree/homol/backend/prisma/manual-copy 
and the last version of migration tool https://github.com/AppCivico/smae/blob/homol/backend/bin/pgsql-migrate.ts

**Reply by ludralph** [reply]
TypedSQL can be used effectively for views, as you mentioned. It allows you to write raw SQL queries for your views and get type-safe results. This can indeed eliminate the need to manually write schemas for views in many cases. For managing complex database objects like functions and triggers, I would recommend using custom SQL migrations. This is mentioned in the [documentation on customizing migrations](https://www.prisma.io/docs/orm/prisma-migrate/workflows/customizing-migrations).

The script you've developed using the pg library seems to be a good approach for managing complex database objects. While there isn't a direct way to integrate this back into Prisma's migration system, you could consider using it alongside Prisma for the parts of your schema that Prisma doesn't handle well.
